File: extensions/sdk-python/nexo_plugin_sdk/manifest.py
"""Minimal TOML reader for `nexo-plugin.toml`.

Validates only the fields the SDK needs at startup (`plugin.id` —
incl. the ASCII-slug regex the host enforces — and `plugin.version`).
The daemon performs full schema validation on the manifest at boot.
"""

import re
from typing import Any

try:
    import tomllib  # Python ≥ 3.11
except ImportError:  # pragma: no cover - 3.10 fallback
    import tomli as tomllib  # type: ignore[no-redef]

from .errors import ManifestError

#: Same shape the host's `nexo-plugin-manifest` crate enforces and
#: the TypeScript SDK validates: `^[a-z][a-z0-9_]{0,31}$`.
PLUGIN_ID_REGEX = re.compile(r"^[a-z][a-z0-9_]{0,31}$")


def read_manifest(toml_text: str) -> dict[str, Any]:
    """Parse manifest TOML, return the full document dict.

    Raises:
        ManifestError: parse failure, missing `plugin.id` /
        `plugin.version`, or a `plugin.id` that violates
        :data:`PLUGIN_ID_REGEX`.
    """
    try:
        data = tomllib.loads(toml_text)
    except Exception as e:  # tomllib.TOMLDecodeError on 3.11+
        raise ManifestError(f"manifest parse failed: {e}") from e

    plugin = data.get("plugin")
    if not isinstance(plugin, dict):
        raise ManifestError("manifest is missing the [plugin] section")

    plugin_id = plugin.get("id")
    if not isinstance(plugin_id, str) or not plugin_id:
        raise ManifestError("manifest is missing required string `plugin.id`")
    if not PLUGIN_ID_REGEX.match(plugin_id):
        raise ManifestError(
            f'plugin.id "{plugin_id}" must match {PLUGIN_ID_REGEX.pattern}'
        )

    version = plugin.get("version")
    if not isinstance(version, str) or not version:
        raise ManifestError("manifest is missing required string `plugin.version`")

    return data

File: extensions/sdk-php/src/Manifest.php
<?php

declare(strict_types=1);

/**
 * Phase 31.5.c — manifest TOML parser.
 *
 * Validates only the fields the SDK needs at construction time.
 * The daemon performs full schema validation at boot, so this
 * stays minimal.
 */

namespace Nexo\Plugin\Sdk;

use Yosymfony\Toml\Toml;

final class Manifest
{
    private const PLUGIN_ID_REGEX = '/^[a-z][a-z0-9_]{0,31}$/';

    /**
     * Parse manifest TOML, return the full document array.
     *
     * @return array<string, mixed>
     * @throws ManifestError
     */
    public static function parse(string $toml): array
    {
        try {
            $raw = Toml::parse($toml);
        } catch (\Throwable $e) {
            throw new ManifestError('manifest TOML parse failed: ' . $e->getMessage());
        }

        if (!is_array($raw)) {
            throw new ManifestError('manifest must parse to a TOML table');
        }

        $plugin = $raw['plugin'] ?? null;
        if (!is_array($plugin)) {
            throw new ManifestError('manifest is missing the [plugin] section', 'plugin');
        }

        $id = self::requireString($plugin, 'id', 'plugin');
        if (preg_match(self::PLUGIN_ID_REGEX, $id) !== 1) {
            throw new ManifestError(
                "plugin.id \"$id\" must match " . self::PLUGIN_ID_REGEX,
                'plugin.id',
            );
        }
        self::requireString($plugin, 'version', 'plugin');
        self::requireString($plugin, 'name', 'plugin');
        self::requireString($plugin, 'description', 'plugin');

        return $raw;
    }

    /**
     * @param array<string, mixed> $obj
     * @throws ManifestError
     */
    private static function requireString(array $obj, string $field, string $ownerLabel): string
    {
        $value = $obj[$field] ?? null;
        if (!is_string($value) || $value === '') {
            throw new ManifestError(
                "$ownerLabel.$field is missing or not a non-empty string",
                "$ownerLabel.$field",
            );
        }
        return $value;
    }
}
